store/index/index.go

```diff
@@ -90,119 +92,226 @@ const BucketPoolSize = 1024
+
+	fi, err := file.Stat()
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
-	if header.BucketsBits != indexSizeBits {
-		return nil, nil, types.ErrIndexWrongBitSize{header.BucketsBits, indexSizeBits}
+
+	idx := &Index{
+		sizeBits:    indexSizeBits,
+		buckets:     buckets,
+		sizeBuckets: sizeBuckets,
+		fileBuckets: fileBuckets,
+		file:        file,
+		fileNum:     lastIndexNum,
+		headerPath:  headerPath,
+		writer:      bufio.NewWriterSize(file, indexBufferSize),
+		Primary:     primary,
+		bucketLk:    sync.RWMutex{},
+		curPool:     make(bucketPool, BucketPoolSize),
+		nextPool:    make(bucketPool, BucketPoolSize),
+		length:      types.Position(fi.Size()),
+		basePath:    path,
+		updateSig:   make(chan struct{}, 1),
+		gcDone:      make(chan struct{}),
 	}
-	buckets, err := NewBuckets(indexSizeBits)
+
+	go idx.garbageCollector()
+
+	return idx, nil
+}
+
+func indexFileName(basePath string, fileNum uint32) string {
+	return fmt.Sprintf("%s.%d", basePath, fileNum)
+}
+
+func scanIndex(basePath string, fileNum uint32, indexSizeBits uint8, buckets Buckets, sizeBuckets SizeBuckets, fileBuckets FileBuckets) (uint32, error) {
+	if len(buckets) != 1<<indexSizeBits {
+		panic("wrong size of buckets")
+	}
+	if len(sizeBuckets) != 1<<indexSizeBits {
+		panic("wrong size of sizeBuckets")
+	}
+	if len(fileBuckets) != 1<<indexSizeBits {
+		panic("wrong size of fileBuckets")
+	}
+
+	var lastFileNum uint32
+	for {
+		err := scanIndexFile(basePath, fileNum, indexSizeBits, buckets, sizeBuckets, fileBuckets)
+		if err != nil {
+			if os.IsNotExist(err) {
+				break
+			}
+			return 0, err
+		}
+		lastFileNum = fileNum
+		fileNum++
+	}
+	return lastFileNum, nil
+}
+
+func (i *Index) StorageSize() (int64, error) {
+	header, err := readHeader(i.headerPath)
 	if err != nil {
-		return nil, nil, err
+		if os.IsNotExist(err) {
+			return 0, nil
+		}
+		return 0, err
 	}
-	sizeBuckets, err := NewSizeBuckets(indexSizeBits)
+	var size int64
+	fi, err := os.Stat(i.headerPath)
 	if err != nil {
-		return nil, nil, err
+		return 0, err
 	}
-	buffered := bufio.NewReader(file)
-	iter := NewIndexIter(buffered, types.Position(bytesRead))
+	size += fi.Size()
+
+	fi, err = os.Stat(i.basePath + ".free")
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return 0, err
+		}
+	} else {
+		size += fi.Size()
+	}
+
+	fileNum := header.FirstFile
 	for {
-		data, pos, done, err := iter.Next()
-		if done {
-			break
+		fi, err = os.Stat(indexFileName(i.basePath, fileNum))
+		if err != nil {
+			if os.IsNotExist(err) {
+				break
+			}
+			return 0, err
 		}
-		if err == io.EOF {
-			// The file is corrupt. Though it's not a problem, just take the data we
-			// are able to use and move on.
-			if _, err := file.Seek(0, 2); err != nil {
-				return nil, nil, err
+		size += fi.Size()
+		fileNum++
+	}
+	return size, nil
+}
+
+func scanIndexFile(basePath string, fileNum uint32, indexSizeBits uint8, buckets Buckets, sizeBuckets SizeBuckets, fileBuckets FileBuckets) error {
```

Comment:
`indexSizeBits` is not used.

Reply:
Oh, look at that.  I moved the allocation of the buckets outside of this function, which is all it was used for.